File: class04/java/TCC0401-AFDTabela.java
import java.util.*;

/**
 * TCC0401-AFDTabela.java | Aula 04
 * Simulador de AFD por tabela de transição.
 * AFD: palavras sobre {a,b} que contêm aa ou bb como subpalavra.
 *
 * Mesma lógica e mesma saída do TCC0401-AFDTabela.py.
 */
class AFDTabela {

    static Map<String, Map<String, String>> transicoes = Map.of(
            "q0", Map.of("a", "q1", "b", "q2"),
            "q1", Map.of("a", "qf", "b", "q2"),
            "q2", Map.of("a", "q1", "b", "qf"),
            "qf", Map.of("a", "qf", "b", "qf")
    );

    static String estadoInicial = "q0";
    static Set<String> estadosFinais = Set.of("qf");

    static boolean aceita(String palavra) {
        String estado = estadoInicial;
        for (char c : palavra.toCharArray()) {
            Map<String, String> trans = transicoes.get(estado);
            if (trans == null) return false;
            estado = trans.get(String.valueOf(c));
            if (estado == null) return false;
        }
        return estadosFinais.contains(estado);
    }

    static void testar(List<String> palavras) {
        System.out.printf("%-12s %-10s %s%n", "Palavra", "Resultado", "Estados");
        System.out.println("-".repeat(55));
        for (String w : palavras) {
            List<String> estados = new ArrayList<>();
            String estado = estadoInicial;
            estados.add(estado);
            for (char s : w.toCharArray()) {
                Map<String, String> trans = transicoes.get(estado);
                if (trans == null) break;
                estado = trans.get(String.valueOf(s));
                if (estado == null) break;
                estados.add(estado);
            }
            String resultado = aceita(w) ? "ACEITA" : "REJEITADA";
            String trace = String.join(" → ", estados);
            System.out.printf("%-12s %-10s %s%n", w, resultado, trace);
        }
    }

    public static void main(String[] args) {
        List<String> palavras = List.of(
                "abba", "aba", "bab", "aabb", "abab",
                "baa", "bba", "aa", "bb", "ab");
        testar(palavras);
    }
}

File: class04/python/TCC0401-AFDTabela.py
#!/usr/bin/env python3
"""TCC0401-AFDTabela.py | Aula 04
Simulador de AFD por tabela de transição.
AFD: palavras sobre {a,b} que contêm aa ou bb como subpalavra.

M = ({a,b}, {q0,q1,q2,qf}, δ, q0, {qf})
"""

transicoes = {
    ("q0", "a"): "q1", ("q0", "b"): "q2",
    ("q1", "a"): "qf", ("q1", "b"): "q2",
    ("q2", "a"): "q1", ("q2", "b"): "qf",
    ("qf", "a"): "qf", ("qf", "b"): "qf",
}

estado_inicial = "q0"
estados_finais = {"qf"}


def aceita(palavra: str) -> bool:
    """Retorna True se a palavra é aceita pelo AFD."""
    estado = estado_inicial
    for simbolo in palavra:
        estado = transicoes.get((estado, simbolo))
        if estado is None:
            return False
    return estado in estados_finais


def testar(palavras):
    print(f"{'Palavra':<12} {'Resultado':<10} {'Estados'}")
    print("-" * 55)
    for w in palavras:
        estados = [estado_inicial]
        estado = estado_inicial
        for s in w:
            estado = transicoes.get((estado, s))
            if estado is None:
                break
            estados.append(estado)
        resultado = "ACEITA" if aceita(w) else "REJEITADA"
        trace = " → ".join(estados)
        print(f"{w:<12} {resultado:<10} {trace}")


if __name__ == "__main__":
    palavras = ["abba", "aba", "bab", "aabb", "abab",
                "baa", "bba", "aa", "bb", "ab"]
    testar(palavras)
